core/http-auth-aws/src/main/java/software/amazon/awssdk/http/auth/aws/util/SignerUtils.java

```diff
@@ -201,30 +186,57 @@ public static void addDateHeader(SdkHttpRequest.Builder requestBuilder, String d
         requestBuilder.putHeader(SignerConstant.X_AMZ_DATE, dateTime);
     }
 
-    public static void addSha256ContentHeader(SdkHttpRequest.Builder requestBuilder, ContentChecksum contentChecksum) {
-        requestBuilder.firstMatchingHeader(SignerConstant.X_AMZ_CONTENT_SHA256)
-            .filter(h -> h.equals("required"))
-            .ifPresent(h ->
-                requestBuilder.putHeader(
-                    SignerConstant.X_AMZ_CONTENT_SHA256, contentChecksum.contentHash()));
+    private static MessageDigest getMessageDigestInstance() {
+        try {
+            return MessageDigest.getInstance("SHA-256");
+        } catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException("Unable to get SHA256 Function: " + e.getMessage());
+        }
     }
 
-    /**
-     * Add a checksum header if the checksum is not null and the payload is signed
-     */
-    public static void addChecksumHeader(SdkHttpRequest.Builder requestBuilder, SdkChecksum sdkChecksum,
-                                         String contentHashString,
-                                         String checksumHeaderName) {
+    public static InputStream getBinaryRequestPayloadStream(ContentStreamProvider streamProvider) {
+        try {
+            if (streamProvider == null) {
+                return new ByteArrayInputStream(new byte[0]);
+            }
+            return streamProvider.newStream();
+        } catch (Exception e) {
+            throw new RuntimeException("Unable to read request payload to sign request: " + e.getMessage());
+        }
+    }
 
-        if (sdkChecksum != null && !isUnsignedPayload(contentHashString)) {
-            requestBuilder.putHeader(checksumHeaderName, BinaryUtils.toBase64(sdkChecksum.getChecksumBytes()));
+    public static byte[] hash(InputStream input) {
+        try {
+            MessageDigest md = getMessageDigestInstance();
```

Comment:
Can we add a TODO: performance test to see if we should cache these message digest instances in a thread local, like we do in the sigv4 signer?